packages/SwingSet/src/kernel/kernel.js

```diff
@@ -323,101 +303,87 @@ export default function buildKernel(
     try {
+      /** @type { VatDeliveryResult } */
       // eslint-disable-next-line no-use-before-define
       const deliveryResult = await vatWarehouse.deliverToVat(vatID, kd, vd, vs);
-      insistVatDeliveryResult(deliveryResult);
-      if (vd[0] !== 'bringOutYourDead') {
-        if (vatKeeper.countdownToReap()) {
-          kernelKeeper.scheduleReap(vatID);
-        }
-      }
-      const [status, problem] = deliveryResult;
-      if (status !== 'ok') {
-        // probably a metering fault, or a bug in the vat's dispatch()
-        console.log(`delivery problem, terminating vat ${vatID}`, problem);
-        setTerminationTrigger(vatID, true, true, makeError(problem));
-        return harden(['crank-failed', {}]);
-      }
 
+      insistVatDeliveryResult(deliveryResult);
+      // const [ ok, problem, usage ] = deliveryResult;
       if (deliveryResult[0] === 'ok') {
-        let used;
-        const metering = deliveryResult[2];
-        if (metering) {
-          // if the result has metering, we report it to the runPolicy
-          const consumed = metering.compute;
-          assert.typeof(consumed, 'number');
-          used = BigInt(consumed);
-          policyInput = ['crank', { computrons: used }];
-          if (useMeter && meterID) {
-            // If we have a Meter and we want to use it, do so.
-            const underflow = deductMeter(meterID, used, true);
-            if (underflow) {
-              console.log(
-                `meter ${meterID} underflow, terminating vat ${vatID}`,
-              );
-              const err = makeError('meter underflow, vat terminated');
-              setTerminationTrigger(vatID, true, true, err);
-              return harden(['crank-failed', {}]);
-            }
-          }
-        }
+        return { metering: deliveryResult[2] };
+      } else {
+        // probably a hard metering fault, or a bug in the vat's dispatch()
+        return { terminate: deliveryResult[1] }; // might be dead
       }
     } catch (e) {
       // log so we get a stack trace
       console.error(`error in kernel.deliver:`, e);
       throw e;
     }
-    return harden(policyInput);
   }
 
   /**
    * Deliver one message to a vat.
    *
-   * @param { string } vatID
+   * @param { VatID } vatID
    * @param { string } target
-   * @param { * } msg
-   * @returns { Promise<PolicyInput> }
+   * @param { Message } msg
+   * @returns { Promise<DeliveryStatus | null> }
    */
-  async function deliverToVat(vatID, target, msg) {
+  async function processSend(vatID, target, msg) {
     insistMessage(msg);
-    /** @type { PolicyInput } */
-    let policyInput = ['none'];
     kernelKeeper.incStat('dispatches');
     kernelKeeper.incStat('dispatchDeliver');
     // eslint-disable-next-line no-use-before-define
-    if (!vatWarehouse.lookup(vatID)) {
-      if (msg.result) {
-        resolveToError(msg.result, VAT_TERMINATION_ERROR);
-      }
-    } else {
+    if (vatWarehouse.lookup(vatID)) {
```

Comment:
Nit: I wish we were consistent in which path we handle exceptional conditions. For example `processNotify` handles errors inside the `if` block, and the normal condition outside of it.

Reply:
Yeah, in general I'm trying to avoid nesting: each `if` clause ends with a `return` so you don't have to worry about what else might happen later in the function. After that, I try to make the initial `if` be a positive case: `if (foo())` rather than `if (!foo())`.

`processSend` has a fairly simple thing to do in the positive case (send to the vat), whereas the negative case has an extra conditional (return `null`, but also reject the result promise iff it has one). If it weren't for the extra typescript/eslint-disable comments, that positive clause would be shorter than the negative clause:

```js
    if (vatWarehouse.lookup(vatID)) {
      const kd = harden(['message', target, msg]);
      const vd = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd);
      return deliverAndLogToVat(vatID, kd, vd);
    }
```

vs

```js
    if (msg.result) {
      resolveToError(msg.result, VAT_TERMINATION_ERROR);
    }
    return null;
```

But for `processNotify`, there are more non-delivery cases to peel off first (dead vat, already-resolved so not in c-list, already-retired so not in promise-table), so doing the negative cases first reduces the nesting.

But.. I like the idea of having the delivery be the last thing in each handler, that seems like the best consistency to aim for. So I think I'll switch `processSend` around:

```js
  async function processSend(vatID, target, msg) {
    insistMessage(msg);
    kernelKeeper.incStat('dispatches');
    kernelKeeper.incStat('dispatchDeliver');
    // eslint-disable-next-line no-use-before-define
    if (!vatWarehouse.lookup(vatID)) {
      // splat
      if (msg.result) {
        resolveToError(msg.result, VAT_TERMINATION_ERROR);
      }
      return null;
    }

    /** @type { KernelDeliveryMessage } */
    const kd = harden(['message', target, msg]);
    // eslint-disable-next-line no-use-before-define
    const vd = vatWarehouse.kernelDeliveryToVatDelivery(vatID, kd);
    return deliverAndLogToVat(vatID, kd, vd);
  }
```